Feature/#22 profile update

dain391 opened the pull request:
## 개요
> 프로필 수정 및 조회 기능 전체 구현 완료
<img width="608" height="187" alt="스크린샷 2025-10-08 18 14 30" src="https://github.com/user-attachments/assets/3420cfeb-af25-43ca-ace1-f0117756d6b6" />
<img width="608" height="187" alt="스크린샷 2025-10-08 18 14 47" src="https://github.com/user-attachments/assets/f7ed5e8d-af64-4eb5-9748-3d1fadfd143f" />


## 작업 사항
- 프로필 수정 API 구현 (Controller, Facade, Service 계층 완성)
- User 엔티티에 사용자 정보 수정 메서드 추가
- ProfileResponse, ProfileUpdateRequest DTO 생성
- SecurityConfig에 permitAll URL 수정
- ProfileService 단위 테스트 및 통합 테스트 추가
- ProfileController 통합 테스트 추가
- ddl-auto update → create 로 변경 (테스트 환경 오류 해결)

## 리뷰 요청 포인트
- 서비스와 파사드 계층 간 책임 분리가 적절한지 확인 부탁드립니다.
- 수정 가능한 필드 검증 로직이 누락된 부분 없는지 확인 부탁드립니다.

## 기타 사항
관련 이슈: #22 
참고 자료:

dain391 committed:
refactor: 프로필 조회 로직 및 DTO 매핑 구조 개선

- ProfileService에서 ProfileResponse.from(User) 사용하도록 변경
- ProfileFacade 단순 위임 구조로 리팩토링 (Service 호출만 수행)
- ProfileResponse.from(User) 정적 팩토리 메서드 구현 (User 엔티티 필드 매핑)
- name, nickname, phoneNumber, isProfilePublic 필드 변환 처리

dain391 committed:
feat: 프로필 수정 요청 DTO 생성

- 이메일, 비밀번호, 이름, 닉네임, 전화번호, 공개여부 필드 추가
- 유효성 검증 어노테이션(@Email, @Size, @Pattern) 적용

#22

dain391 committed:
feat: 프로필 수정 서비스 로직 구현

- 이메일, 비밀번호, 이름, 닉네임, 전화번호, 공개여부 수정 가능하도록 구현
- 비밀번호 변경 시 PasswordEncoder 적용
- User.updateProfile() 호출을 통한 엔티티 필드 반영

#22

dain391 committed:
feat: 프로필 수정 Facade 계층 구현

- ProfileService를 호출하여 프로필 수정 기능 수행
- 트랜잭션(readOnly, write) 분리 적용
- Controller에서 비즈니스 로직을 분리하고 흐름 제어 역할 수행

#22

dain391 committed:
feat: ProfileController 조회 및 수정 API 구현

- 이메일, 비밀번호, 이름, 닉네임, 전화번호, 공개여부 수정 기능 추가
- @AuthenticationPrincipal로 인증 사용자 ID 주입
- @Valid 기반 요청 DTO 검증 적용
- ApiResponse 표준 응답 구조 통일

#22

dain391 committed:
feat: User 엔티티 사용자 정보 수정 메서드 구현

- updateProfile 메서드 구현
- 공개 여부 변경 시 profilePublicAt 자동 갱신 처리
- null 또는 공백 입력 시 기존 값 유지

#22

dain391 committed:
fix: SecurityConfig permit URL 수정

dain391 committed:
test: 프로필 수정 API 테스트 코드 추가

- 프로필 수정(성공/비로그인 사용자 테스트 추가)
- PATCH /api/profile 요청에 대한 인증 실패(401) 검증 포함
- MockMvc 사용해 requestBody JSON 및 Authorization 헤더 테스트

#22

dain391 committed:
feat: ddl-auto update -> create로 변경

#22

dain391 committed:
test: ProfileService 단위 테스트 추가 및 프로필 수정 로직 검증

- ProfileServiceTest 클래스 작성
  - 프로필 조회 성공/실패 테스트 추가
  - 프로필 수정 성공/실패 테스트 추가
- ProfileResponse에 email 필드 추가
- Mockito 기반 Mock Repository/Encoder 설정 및 검증

#22

dain391 committed:
Merge branch 'dev' into feature/#22-profile-update

dain391 committed:
chore: 프로필 API 경로를 /api/profile → /v1/profile 로 변경

#22

dain391 committed:
style: ProfileController 불필요한 임포트 제거

#22

Hojeong016 commented on src/main/java/com/babgo/controller/profile/dto/ProfileUpdateRequest.java:
```diff
@@ -0,0 +1,31 @@
+package com.babgo.controller.profile.dto;
+
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Pattern;
+import jakarta.validation.constraints.Size;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+@Getter
+@NoArgsConstructor
+public class ProfileUpdateRequest {
+
+    @Email(message = "이메일 형식이 올바르지 않습니다.")
+    private String email;
+
+    @Size(min = 8, message = "비밀번호는 8자 이상이어야 합니다.")
+    private String password;
+
+    @NotBlank(message = "이름은 필수 입력 항목입니다.")
+    @Size(min = 2, max = 20, message = "이름은 2자 이상이어야 합니다.")
+    private String name;
+
+    @Size(min = 2, max = 20, message = "닉네임은 2자 이상, 20자 이하여야 합니다.")
+    private String nickname;
+
+    @Pattern(regexp = "^010\\d{8}$", message = "전화번호 형식이 올바르지 않습니다.")
+    private String phoneNumber;
+
+    private Boolean isProfilePublic;
+}
```
DTO 불변성 확인 부탁드립니다!

dain391 replied:
모든 필드를 `final`로 변경하고, `@NoArgsConstructor` 대신 생성자를 통해 값이 주입되도록 수정했습니다.

Hojeong016 commented on src/main/java/com/babgo/domain/profile/ProfileService.java:
```diff
@@ -1,49 +1,52 @@
-        user.updateProfilePublic(isProfilePublic);
-
-        log.info("프로필 업데이트 완료: userId={}, nickname={}", user.getUserId(), nickname);
+        return ProfileResponse.from(user);
```
해당 dto 는 controller에서 클라이언트 통신용 DTO 로 해석을 하였는데 
올바르게 이해한게 맞을까요?

그렇다면 
해당 도메인 서비스에서 controller 계층을 직접 의존하게 되는 방향이 되는 것 같습니다. 
중간 dto를 사용해보시는 건 어떠신가요?

dain391 replied:
맞습니다! `ProfileUpdateRequest`는 Controller ↔ 클라이언트 통신용 DTO로 사용하고 있습니다.  
리뷰 주신 대로 Service 계층이 직접 의존하지 않도록 `ProfileInfo`라는 중간 DTO를 추가하였습니다.  
현재 `Facade`에서 `ProfileInfo.from(request)`로 변환한 뒤, `ProfileService`는 `ProfileInfo`만 의존하도록 계층 분리 완료했습니다.

Hojeong016 reviewed (commented):
수고하셨습니다! 디코 알람이 밀려서 늦게 확인했네요 죄송합니다🥹

dain391 committed:
refactor: ProfileResponse에 빌더 제거 및 팩토리 메서드 적용

#22

dain391 committed:
refactor: ProfileUpdateRequest 불변성 확보 및 ProfileInfo 중간 DTO 추가

- ProfileService 계층 분리를 위한 ProfileInfo 적용 및 DTO 의존성 제거
- ProfileServiceTest 수정

#22

Hojeong016 reviewed (approved):
LGTM 👍
